pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufRecordReader.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.pinot.spi.data.readers.RecordReaderConfig;
+import org.apache.pinot.spi.data.readers.RecordReaderUtils;
+import org.apache.pinot.spi.utils.ResourceFinder;
+import org.apache.pinot.spi.utils.SchemaFieldExtractorUtils;
+
+
+public class ProtoBufRecordReader implements RecordReader {
+  private File _dataFile;
+  private Schema _schema;
+  private ProtoBufRecordExtractor _recordExtractor;
+
+  private InputStream _inputStream;
+  private boolean _hasNext;
+  private Descriptors.Descriptor _descriptor;
+
+  private boolean hasMoreToRead()
+      throws IOException {
+    _inputStream.mark(1);
+    int nextByte = _inputStream.read();
+    _inputStream.reset();
+    return nextByte != -1;
+  }
+
+  private void init()
+      throws IOException {
+    _inputStream = RecordReaderUtils.getBufferedInputStream(_dataFile);
+    try {
+      _hasNext = hasMoreToRead();
+    } catch (Exception e) {
+      _inputStream.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public void init(File dataFile, Schema schema, @Nullable RecordReaderConfig recordReaderConfig)
+      throws IOException {
+    _dataFile = dataFile;
+    _schema = schema;
+    Set<String> sourceFields = SchemaFieldExtractorUtils.extract(schema);
+    ProtoBufRecordExtractorConfig recordExtractorConfig = new ProtoBufRecordExtractorConfig();
+    ProtoBufRecordReaderConfig protoBufRecordReaderConfig = (ProtoBufRecordReaderConfig) recordReaderConfig;
+    InputStream fin = getDescriptorFileInputStream(protoBufRecordReaderConfig);
+    buildProtoBufDescriptor(fin);
+    _recordExtractor = new ProtoBufRecordExtractor();
+    _recordExtractor.init(sourceFields, recordExtractorConfig);
+    init();
+  }
+
+  private void buildProtoBufDescriptor(InputStream fin)
+      throws IOException {
+    try {
+      DescriptorProtos.FileDescriptorSet set = DescriptorProtos.FileDescriptorSet.parseFrom(fin);
+      Descriptors.FileDescriptor fileDescriptor =
+          Descriptors.FileDescriptor.buildFrom(set.getFile(0), new Descriptors.FileDescriptor[]{});
+      _descriptor = fileDescriptor.getMessageTypes().get(0);
+    } catch (Descriptors.DescriptorValidationException e) {
+      throw new IOException("Descriptor file validation failed", e);
+    }
+  }
+
+  private InputStream getDescriptorFileInputStream(ProtoBufRecordReaderConfig protoBufRecordReaderConfig)
+      throws IOException {
+    URI descriptorFileURI = protoBufRecordReaderConfig.getDescriptorFile();
+    return ResourceFinder.openResource(descriptorFileURI);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return _hasNext;
+  }
+
+  @Override
+  public GenericRow next()
+      throws IOException {
+    return next(new GenericRow());
+  }
+
+  @Override
+  public GenericRow next(GenericRow reuse)
+      throws IOException {
+    Message message = null;
+    try {
+      DynamicMessage tmp = DynamicMessage.getDefaultInstance(_descriptor);
+      Message.Builder builder = tmp.newBuilderForType();
```

Comment:
Can DynamicMessage and Builder instance also be reused across every call to next?

Reply:
not sure.

Reply:
can you check with the help of your test? If they can be reused, make them class fields?

Reply:
Done.  Temp DynamicMessage can be reused but reusing builder is not possible.
Moved the temp Dyanmic message to class variable.